libtextsecure/message_receiver.js

```diff
@@ -1115,155 +1195,253 @@ MessageReceiver.prototype.extend({
 
     await conversation.setLokiProfile(newProfile);
   },
-  handleDataMessage(envelope, msg) {
+
+  async handleMediumGroupUpdate(envelope, groupUpdate) {
+    const {
+      groupId,
+      groupSecretKey,
+      senderKey,
+      members,
+      groupName,
+    } = groupUpdate;
+
+    const convoExists = window.ConversationController.get(groupId, 'group');
+
+    if (convoExists) {
+      // If the group already exists, check that `members` is empty,
+      // and if so, it is sender key message
+
+      // TODO: introduce TYPE into this message instead?
+      if (!members || !members.length) {
+        log.info('[sender key] got a new sender key from:', envelope.source);
+
+        // We probably don't need to await here
+        await window.SenderKeyAPI.saveSenderKeys(
+          groupId,
+          envelope.source,
+          senderKey
+        );
+
+        this.removeFromCache(envelope);
+        return;
+      }
+
+      log.error(`Conversation for groupId ${groupId} already exists`);
+    }
+
+    const convo = await window.ConversationController.getOrCreateAndWait(
+      groupId,
+      'group'
+    );
+    convo.set('is_medium_group', true);
+    convo.set('active_at', Date.now());
+    convo.set('name', groupName);
+
+    await window.Signal.Data.createOrUpdateIdentityKey({
+      id: groupId,
+      secretKey: groupSecretKey,
+    });
+
+    // Save sender's key
+    await window.SenderKeyAPI.saveSenderKeys(
+      groupId,
+      envelope.source,
+      senderKey
+    );
+
+    // TODO: Check that we are even a part of this group?
+    const ourIdentity = await textsecure.storage.user.getNumber();
+
+    const ownSenderKey = await window.SenderKeyAPI.createSenderKeyForGroup(
+      groupId,
+      ourIdentity
+    );
+
+    {
+      // TODO: Send own key to every member
+
+      const otherMembers = _.without(members, ourIdentity);
+
+      const proto = new textsecure.protobuf.DataMessage();
+
+      // We reuse the same message type for sender keys
+      const update = new textsecure.protobuf.MediumGroupUpdate();
+      update.groupId = groupId;
+      update.senderKey = ownSenderKey;
+
+      proto.mediumGroupUpdate = update;
+
+      // TODO: send to our linked devices too?
```

Comment:
a linked device should receive this message too normally

Reply:
Correct. I removed the comment already in the upcoming PR.